Cloudkit

drewmccormack opened the pull request.

drewmccormack committed:
Added method that can fetch changes in the PARStore only from the local device.

drewmccormack committed:
Merge branch 'master' of github.com:cparnot/PARStore into cloudkit

drewmccormack committed:
Added method for inserting data from remote devices into DBs in PARStore.

This is needed to support cloud services like CloudKit, which have no local storage.

Also set undoManager on all MOCs to nil in PARStore. Otherwise is wasting time and memory.

drewmccormack committed:
Added a createDirectory to addChanges… method.

This makes sure the device directory exists when a new device joins remotely.

drewmccormack committed:
Added a check for the timestamp of new changes in addChanges…

Changed the method name to ‘appendChanges…’ to make clear we can only add changes to the end of the log.

We now fetch the existing maximum timestamp for a given device, and if a new change has a timestamp smaller, it is ignored.

drewmccormack committed:
Fixed the NSExpression used to get the maximum timestamp.

drewmccormack committed:
Main context now merges background changes.

Added observation of `NSManagedObjectContextDidSaveNotification`.

drewmccormack committed:
Replaced the fetchChangesSinceTimestamp:includeOnlyLocalChanges: with a fetch for any given device.

We need to be able to fetch per device now with CloudKit. This new method is more flexible.

drewmccormack committed:
Added a method that can fetch changes between two timestamps.

We need this for the new change set processing.

alexandergriekspoor commented on Core/PARStore.h:
```diff
@@ -95,17 +96,41 @@ extern NSString *PARStoreDidSyncNotification;
+/// This method returns an array of PARChange instances for the device identifier passed in.
+/// It should not be called from within a transaction, or it will fail.
+/// Pass in nil for the device identifier to get results for all devices.
+- (NSArray *)fetchChangesSinceTimestamp:(nullable NSNumber *)timestampLimit forDeviceIdentifier:(nullable NSString *)deviceID;
```
I noticed the discrepancy between `timestamp` above and `timestampLimit` here, but then also noticed that in the .m file the argument is called `timestampLimit` in both cases. My feeling says it should be just `timestamp` in both cases and also in both .m and .h file, but not very attached as long as they're the same for both methods.

drewmccormack replied:
Yeah, copy-and-paste issues. Using 'timestamp' now.

alexandergriekspoor commented on Core/PARStore.h:
```diff
@@ -95,17 +96,41 @@ extern NSString *PARStoreDidSyncNotification;
+/// It should not be called from within a transaction, or it will fail.
+/// Pass in nil for the device identifier to get results for all devices.
+/// Pass nil for either timestamp to have an open range.
+- (NSArray *)fetchChangesFromTimestamp:(nullable NSNumber *)firstTimestamp toTimestamp:(nullable NSNumber *)lastTimestamp forDeviceIdentifier:(nullable NSString *)deviceID;
```
similar as above, let's settle on either `deviceID` (used here) or `deviceIdentifier` (used in `appendChanges:` below). My preference would be the latter I guess

drewmccormack replied:
Yes, I'm afraid I've been a bit in limbo with this naming. Waiting for the global change. But going forward, I will try to go with 'identifier' written full out.

alexandergriekspoor replied:
Yeah, I on purposely didn't comment on all the places where I could see the word `uuid` anywhere as I knew I would revisit those anyway when doing the big rename. I just highlighted the ones I would be afraid to miss otherwise. Outside the ones I mentioned don't bother about changing any in what you've written already and indeed just focus on the going forward.

alexandergriekspoor commented on Core/PARStore.h:
```diff
@@ -95,17 +96,41 @@ extern NSString *PARStoreDidSyncNotification;
+
 @interface PARChange : NSObject
 + (PARChange *)changeWithTimestamp:(NSNumber *)timestamp parentTimestamp:(nullable NSNumber *)parentTimestamp key:(NSString *)key propertyList:(id)propertyList;
++ (PARChange *)changeWithPropertyListRepresentation:(id)propertyListRepresentation;
```
Elsewhere in Momenta we call `propertyListRepresentation` a `propertyDictionary`, which IMO already better defines what you expect this `id` to be (a Dictionary of properties)

alexandergriekspoor replied:
`propertyDictionary` also allows for less confusion with the already existing `propertyList` property on `PARChange` as well btw

drewmccormack replied:
Made this change. (Technically, PARStore is not Momenta, so the PropertyDictionary naming doesn't really apply, but still worth changing I think)

alexandergriekspoor commented on Core/PARStore.m:
```diff
@@ -949,6 +950,85 @@ - (void)setEntriesFromDictionary:(NSDictionary *)dictionary
+    NSPersistentStoreCoordinator *psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];
+    
+    // Persistent Store
+    NSError *error = nil;
```
Doesn't seem like we do anything with this error?

drewmccormack replied:
Added property error handling now. The method returns an error.

alexandergriekspoor commented on Core/PARStore.m:
```diff
@@ -949,6 +950,85 @@ - (void)setEntriesFromDictionary:(NSDictionary *)dictionary
+    [moc setUndoManager:nil];
+
+    // Have main context observe changes made in background, and merge them.
+    id observer = [[NSNotificationCenter defaultCenter] addObserverForName:NSManagedObjectContextDidSaveNotification object:moc queue:nil usingBlock:^(NSNotification *notification) {
```
I've been bitten so often and so badly by how crappy the block based notifications have been implemented by apple that I stay away from them as far as I can, I simply promised myself to at least myself never use them again. If you're happy with them you can leave it, but if you can just use regular ones I'd be more happy with those.

drewmccormack replied:
They work as advertised. You just have to follow the rules, which is to remove the returned observer object. We are doing that, and it avoids having to add extra methods for what is quite a localised usage, so I think it is OK here.

alexandergriekspoor commented on Core/PARStore.m:
```diff
@@ -949,6 +950,85 @@ - (void)setEntriesFromDictionary:(NSDictionary *)dictionary
+    [moc performBlockAndWait:^{
+        NSError *error = nil;
+        
+        // Expression to retrieve the maximum timestamp in the store
```
I thought we had discussed that we cannot rely on this in order to make a robust system where changes might be added not in order?

My initial thought when seeing the expression based stuff is, why --if we any way need to rely on order of insertion-- are we not simply fetching the last row in the db?

However, I thought we discussed how we will need to be robust against:
a) receiving batches of changes out of order, and
b) be resilient against batches containing the overlapping changes

Both of these you would (have to) address by checking for each change if the row is in the db already or not, more expensive but most reliable. I guess that would been querying for timestamp + parenttimestamp + key.

drewmccormack replied:
I can certainly change this method so that it uses queries. I agree that would be more robust.

My main concern is that given that PARStore assumes append only for each device, I think it may require far reaching changes to PARStore and the revision manager if changes are able to be inserted at any point in the history. (Reloading the whole document history every sync is not an option IMO.)

I will see if I can make the change, but it does not solve the issue of potentially inserting into the history of PARStore and breaking the append-only assumption.

When you ask 'why not get the last row'? A database is not ordered, so I would presumably need to do some fetch anyway, with sorting. That would be slower than the @max function I used. But if you have a better idea, let me know. (BTW I don't want to just access cached timestamps, because this activity is happening on a different thread, and I don't want race conditions. Querying the DB itself is safest.)

alexandergriekspoor replied:
My main concern had to do with the fact we have overlapping change sets and we don't want to reinsert the same row twice, the timestamp implementation (without checking the property) is not suitable for this (the querying without checking the data blob isn't 100% either, but much safe, i.e. need to have at least 3 changes of the same property with the same timestamp before getting into trouble).

Re: timestamp, if we indeed cannot rely on the last row inserted is what we need, than @max is the right choice (however moot as we need to move to querying).

drewmccormack replied:
OK, will add querying approach, and just assume that our order of insertion etc takes care of the append-only concern I had.

alexandergriekspoor commented on Core/PARStore.m:
```diff
@@ -2126,64 +2208,110 @@ - (NSNumber *)mostRecentTimestampForKey:(NSString *)key
+         }
+         else {
+             // Filter stores to find one that matches the device.
+             NSArray *eligibleStores = [self.readonlyDatabases filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^(NSPersistentStore *store, NSDictionary *bindings) {
```
Wow, looking at this line after having been in the swift world for a while really does make objc look archaic!

drewmccormack replied:
Not ObjC, just the predicate API. Most devs have their own block based filter method, which is much more readable.

Have changed this to make it more legible.

alexandergriekspoor replied:
Ah apologies, I didn't want to suggest any changes, was more a generic remark

alexandergriekspoor reviewed (changes requested):
Some comments, but major issue is around `appendChanges:forDeviceIdentifier:` not relying on the latest timestamp but really checking if the row already exists or not using one query per change.

drewmccormack committed:
Changes requested in pull request.

drewmccormack committed:
Changed the appendChanges: method

It is now an insertChanges: method. It can insert at any position, or it can do an append only insert. Importantly, in both cases, it now checks for duplicates, and will allow a new change with the same timestamp as the existing maximum timestamp (provided the new change is unique).

drewmccormack commented:
Updated the appendChanges method as requested. Ready for new review.

drewmccormack committed:
Added proper hash method to PARChange.

We don’t want it based on object identity, but on content like key and timestamp. That way, we can use sets to quickly identify if a change is contained in the set, even if the change object is different.

alexandergriekspoor reviewed (approved):
Looking good, over to @cparnot
